common/nodejs-utils/src/distribution.rs

```diff
@@ -0,0 +1,183 @@
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "Node.js" | "node" => Ok(Self::Node),
+            "Yarn" | "yarn" => Ok(Self::Yarn),
+            other => Err(anyhow!("Unknown distribution: {other}")),
+        }
+    }
+}
+
+impl fmt::Display for Distribution {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Node => write!(f, "Node.js"),
+            Self::Yarn => write!(f, "Yarn"),
+        }
+    }
+}
+
+impl Distribution {
+    /// List all versions of a distribution from it's upstream/source location.
+    ///
+    /// # Errors
+    /// - http call failures
+    /// - http response parsing errors
+    pub fn upstream_versions(&self) -> anyhow::Result<VersionSet> {
+        match self {
+            Self::Node => list_upstream_node_versions(),
+            Self::Yarn => list_upstream_yarn_versions(),
+        }
+    }
+
+    /// List all versions of a distribution from it's mirrored location.
+    ///
+    /// # Errors
+    /// - http call failures
+    /// - http response parsing errors
+    pub fn mirrored_versions(&self, bucket: &str) -> anyhow::Result<VersionSet> {
+        s3::list_objects(bucket, DEFAULT_REGION, self.bucket_prefix())?.try_into()
+    }
+
+    /// Build an inventory from the releases listed in the mirrored location.
+    ///
+    /// # Errors
+    /// - http call failures
+    /// - http response parsing errors
+    pub fn mirrored_inventory(&self, bucket: &str) -> anyhow::Result<Inventory> {
+        s3::list_objects(bucket, DEFAULT_REGION, self.bucket_prefix())?.try_into()
+    }
+
+    /// Filter inactive distribution versions and return as a `VersionSet`.
+    ///
+    /// # Errors
+    /// - Active semver range for the distribution is a valid version.
+    pub fn filter_inactive_versions<'i, I>(&self, iter: I) -> anyhow::Result<VersionSet>
+    where
+        I: Iterator<Item = &'i Version>,
+    {
+        let req = self.active_requirement()?;
+        Ok(iter.filter(|v| req.satisfies(v)).cloned().collect())
+    }
+
+    fn bucket_prefix(&self) -> &str {
+        match self {
+            Self::Node => "node",
+            Self::Yarn => "yarn",
+        }
+    }
+    fn active_requirement(self) -> anyhow::Result<Requirement> {
+        Requirement::parse(match self {
+            Self::Node => ">= 16",
+            Self::Yarn => ">= 1.22",
+        })
+        .map_err(|e| anyhow!("{e}"))
+    }
+
+    pub(crate) fn mirrored_path_regex(self) -> anyhow::Result<Regex> {
+        Regex::new(match self {
+            Self::Node => r"node/(?P<channel>\w+)/(?P<arch>[\w-]+)/node-v(?P<version>\d+\.\d+\.\d+)[\w-]+\.tar\.gz",
+            Self::Yarn => r"yarn/(?P<channel>\w+)/yarn-v(?P<version>\d+\.\d+\.\d+(-[\w\.]+)?)\.tar\.gz",
+        }).map_err(|e| anyhow!("Mirrored release regex error: {e}"))
+    }
+}
+
+fn list_upstream_node_versions() -> anyhow::Result<VersionSet> {
+    nodejs_org::list_releases()?
+        .iter()
```

Comment:
Nit, not blocking at all: This could be `into_iter`.

Reply:
Why is `into_iter()` better here? It consumes the vector?

Reply:
Yeah, there is no need to create a "view" to the data when the data will be discarded after the function anyways. With `into_iter` we can work directly with the data, not references to it. In some cases (not here) this can reduce the amount of cloning you need to do in iterator chains.